lib/logstash/outputs/tcp.rb

```diff
@@ -104,50 +109,73 @@ def setup_ssl
+      end
+      # mapping 'TLSv1.2' -> OpenSSL::SSL::OP_NO_TLSv1_2
+      disabled_protocols.map! { |v| OpenSSL::SSL.const_get "OP_NO_#{v.sub('.', '_')}" }
+      @ssl_context.options = disabled_protocols.inject(@ssl_context.options) { |options, op_no| options | op_no }
+    end
+    @ssl_context
+  end
+  private :setup_ssl
+
+  # @note to be able to hook up into #ssl_context from tests
+  def new_ssl_context
+    OpenSSL::SSL::SSLContext.new
+  end
+  private :new_ssl_context
+
+  # @overload Base#register
   def register
     require "socket"
     require "stud/try"
-    if @ssl_enable
-      setup_ssl
-    end # @ssl_enable
+    @closed = Concurrent::AtomicBoolean.new(false)
+    setup_ssl if @ssl_enable
 
     if server?
       @logger.info("Starting tcp output listener", :address => "#{@host}:#{@port}")
       begin
         @server_socket = TCPServer.new(@host, @port)
       rescue Errno::EADDRINUSE
-        @logger.error("Could not start TCP server: Address in use",
-                      :host => @host, :port => @port)
+        @logger.error("Could not start tcp server: Address in use", host: @host, port: @port)
         raise
       end
       if @ssl_enable
         @server_socket = OpenSSL::SSL::SSLServer.new(@server_socket, @ssl_context)
       end # @ssl_enable
-      @client_threads = []
+      @client_threads = Concurrent::Array.new
 
       @accept_thread = Thread.new(@server_socket) do |server_socket|
+        LogStash::Util.set_thread_name("[#{id}]|output|tcp|server_accept")
         loop do
-          Thread.start(server_socket.accept) do |client_socket|
+          break if @closed.value
+          client_socket = server_socket.accept_nonblock exception: false
+          if client_socket == :wait_readable
+            IO.select [ server_socket ]
+            next
+          end
+          Thread.start(client_socket) do |client_socket|
             # monkeypatch a 'peer' method onto the socket.
             client_socket.instance_eval { class << self; include ::LogStash::Util::SocketPeer end }
-            @logger.debug("Accepted connection", :client => client_socket.peer,
-                          :server => "#{@host}:#{@port}")
+            @logger.debug("accepted connection", client: client_socket.peer, server: "#{@host}:#{@port}")
             client = Client.new(client_socket, @logger)
             Thread.current[:client] = client
+            LogStash::Util.set_thread_name("[#{id}]|output|tcp|client_socket-#{@client_threads.size}")
             @client_threads << Thread.current
-            client.run
+            client.run unless @closed.value
           end
         end
       end
 
       @codec.on_event do |event, payload|
-        @client_threads.each do |client_thread|
+        @client_threads.select(&:alive?).each do |client_thread|
           client_thread[:client].write(payload)
         end
-        @client_threads.reject! {|t| !t.alive? }
```

Comment:
Seems like this line was here to cleanup dead client threads, otherwise this array would grow quite large over long executions, maybe I'm missing something but I don't see a replacement for this elsewhere.

Reply:
that was kind of "by design" - with the synchronized array to have only one locking operation (`@client_threads.select(&:alive?)` returns a regular `Array`)
looking at the client loop we break upon an error from `@socket.write` - thus it happens more likely than I have assumed. good find, thanks for checking - updated.